src/module_4/notebook.md

```diff
@@ -0,0 +1,2102 @@
+
+        val_proba = rf.predict_proba(X_val[train_cols])[:, 1]
+        auc_scores_val = auc_score(y_val, val_proba)
+        print(f"Estimators {n}, depth {d}. Validation AUC: {auc_scores_val}")
+
+```
+
+    Estimators 100, depth 9. Train AUC: (0.8431308542157552, 0.22843724976492064)
+    Estimators 100, depth 9. Validation AUC: (0.841019723103303, 0.2004329237782617)
+    Estimators 100, depth 11. Train AUC: (0.8510575200728553, 0.25673821728420076)
+    Estimators 100, depth 11. Validation AUC: (0.8443984795202771, 0.19989639803276882)
+    Estimators 100, depth 13. Train AUC: (0.8631656258184507, 0.29411427996351847)
+    Estimators 100, depth 13. Validation AUC: (0.8437744722414929, 0.19822915066240693)
+    Estimators 150, depth 9. Train AUC: (0.8432346790365738, 0.22881234769151487)
+    Estimators 150, depth 9. Validation AUC: (0.8410070722695171, 0.2005685047376905)
+    Estimators 150, depth 11. Train AUC: (0.851501087061485, 0.25704765292295273)
+    Estimators 150, depth 11. Validation AUC: (0.8444610396081079, 0.20021036219592903)
+    Estimators 150, depth 13. Train AUC: (0.8638383159973727, 0.2951477760867607)
+    Estimators 150, depth 13. Validation AUC: (0.8441732659744031, 0.19837703347620753)
+
+
+The improvements we get by increasing the number of trees and depth are not very significative, thus not worth it the increase in execution time. We stick with `n_estimator=100` and `max_depth=9`.
+
+
+```python
+fig1, ax1 = plt.subplots(1, 2, figsize=(12, 5))
+fig1.suptitle("Train metrics")
+
+fig2, ax2 = plt.subplots(1, 2, figsize=(12, 5))
+fig2.suptitle("Validation metrics")
+
+rf = RandomForestClassifier(n_estimators=100, max_depth=9, random_state=0)
+rf.fit(X_train[train_cols], y_train)
+train_proba = rf.predict_proba(X_train[train_cols])[:, 1]
+plot_roc_and_pr_curves("RF", y_pred=train_proba, y_test=y_train, figure=(fig1, ax1))
+
+val_proba = rf.predict_proba(X_val[train_cols])[:, 1]
+plot_roc_and_pr_curves("RF", y_pred=val_proba, y_test=y_val, figure=(fig2, ax2))
+
+coeffs = rf.feature_importances_
+feat_importance = pd.DataFrame({"feature": train_cols, "importance": coeffs})
+feat_importance = feat_importance.sort_values('importance', ascending=True)
+feat_importance.plot(kind='barh', x='feature', y='importance', title="Feature Importance", figsize=(9, 5))
+        
+plot_roc_and_pr_curves("Baseline", y_pred=val_df["global_popularity"], y_test=val_df[label_col], figure=(fig2, ax2))
+```
+
+
+    
+![png](notebook_files/notebook_32_0.png)
+    
+
+
+
+    
+![png](notebook_files/notebook_32_1.png)
+    
+
+
+
+    
+![png](notebook_files/notebook_32_2.png)
+    
+
+
+### Gradient Boosting
+
+
+```python
+roc_scores_train = []
+roc_scores_val = []
+pr_scores_train = []
+pr_scores_val = []
+
+n_estimtors = [1, 10, 50, 100]
+max_depths = [2, 5, 7, 9]
+learnning_rate = [0.001, 0.01, 0.1]
+
+hyper_estimators = []
+hyper_depths = []
+hyper_lr = []
+
+for n in n_estimtors:
```

Comment:
You can use: https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.ParameterGrid.html to make these loops more elegant